drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+	if (ret) {
+		DRV_LOG(ERR, "%s fail to get notify area", vdev->device->name);
+		return ret;
+	}
+
+	DRV_LOG(DEBUG, "Vid %d qid:%d offset:0x%"PRIx64"size:0x%"PRIx64, vid, qid,*offset,*size);
+	return 0;
+}
+
+static struct rte_vdpa_dev_ops virtio_vdpa_ops = {
+	.get_queue_num = virtio_vdpa_vqs_max_get,
+	.get_features = virtio_vdpa_features_get,
+	.get_protocol_features = virtio_vdpa_protocol_features_get,
+	.dev_conf = virtio_vdpa_dev_config,
+	.dev_close = virtio_vdpa_dev_close,
+	.set_vring_state = virtio_vdpa_vring_state_set,
+	.set_features = virtio_vdpa_features_set,
+	.migration_done = NULL,
+	.get_vfio_group_fd = virtio_vdpa_group_fd_get,
+	.get_vfio_device_fd = virtio_vdpa_device_fd_get,
+	.get_notify_area = virtio_vdpa_notify_area_get,
+	.get_stats_names = NULL,
+	.get_stats = NULL,
+	.reset_stats = NULL,
+};
+
+static int vdpa_check_handler(__rte_unused const char *key,
+		const char *value, void *ret_val)
+{
+	if (strcmp(value, "1") == 0)
+		*(int *)ret_val = 1;
+	else
+		*(int *)ret_val = 0;
+
+	return 0;
+}
+
+#define VIRTIO_ARG_VDPA 	  "vdpa"
+
+static int
+virtio_pci_devargs_parse(struct rte_devargs *devargs, int *vdpa)
+{
+	struct rte_kvargs *kvlist;
+	int ret = 0;
+
+	if (devargs == NULL)
+		return 0;
+
+	kvlist = rte_kvargs_parse(devargs->args, NULL);
+	if (kvlist == NULL) {
+		DRV_LOG(ERR, "Error when parsing param");
+		return 0;
+	}
+
+	if (rte_kvargs_count(kvlist, VIRTIO_ARG_VDPA) == 1) {
+		/* vdpa mode selected when there's a key-value pair:
+		 * vdpa=1
+		 */
+		ret = rte_kvargs_process(kvlist, VIRTIO_ARG_VDPA,
+				vdpa_check_handler, vdpa);
+		if (ret < 0)
+			DRV_LOG(ERR, "Failed to parse %s", VIRTIO_ARG_VDPA);
+	}
+
+	rte_kvargs_free(kvlist);
+
+	return ret;
+}
+
+static void
+virtio_vdpa_queues_free(struct virtio_vdpa_priv *priv)
+{
+	uint16_t nr_vq = priv->hw_nr_virtqs;
+	struct virtio_vdpa_vring_info *vr;
+	uint16_t i;
+
+	if(priv->vrings) {
+		for (i = 0; i < nr_vq; i++) {
+			vr = priv->vrings[i];
+			if (!vr)
+				continue;
+			rte_free(vr);
+			priv->vrings[i] = NULL;
+		}
+		rte_free(priv->vrings);
+		priv->vrings = NULL;
+	}
+
+	virtio_pci_dev_queues_free(priv->vpdev,nr_vq);
+
+}
+
+static int
+virtio_vdpa_queues_alloc(struct virtio_vdpa_priv *priv)
+{
+	uint16_t nr_vq = priv->hw_nr_virtqs;
+	struct virtio_vdpa_vring_info *vr;
+	uint16_t i;
+	int ret;
+
+	ret = virtio_pci_dev_queues_alloc(priv->vpdev, nr_vq);
+	if (ret) {
+		DRV_LOG(ERR, "Failed to alloc virtio device queues.");
+		return ret;
+	}
+
+	priv->vrings = rte_zmalloc(NULL, sizeof(struct virtio_vdpa_vring_info *) * nr_vq, 0);
+	if (!priv->vrings) {
+		virtio_vdpa_queues_free(priv);
+		return -ENOMEM;
+	}
+
+	for (i = 0; i < nr_vq; i++) {
+		vr = rte_zmalloc_socket(NULL, sizeof(struct virtio_vdpa_vring_info), RTE_CACHE_LINE_SIZE,
+					priv->pdev->device.numa_node);
+		if (vr == NULL) {
+			virtio_vdpa_queues_free(priv);
+			return -ENOMEM;
+		}
+		priv->vrings[i] = vr;
+		priv->vrings[i]->index = i;
+		priv->vrings[i]->priv = priv;
```

Comment:
lets double check if for all 256 vqs do we need to store this pointer or we can get the priv from the device itself.